blocks/api/parser.js

```diff
@@ -123,13 +123,54 @@ export function isValidBlock( rawContent, blockType, attributes ) {
 
 	const isValid = ( actual === expected );
 
-	if ( ! isValid && 'development' === process.env.NODE_ENV ) {
-		// eslint-disable-next-line no-console
-		console.error(
-			'Invalid block parse\n' +
-				'\tExpected: ' + expected + '\n' +
-				'\tActual:   ' + actual
-		);
+	if ( ! isValid ) {
+		// NOTE: DO NOT COLLAPSE THESE CONDITIONS OR CONVERT COMMONJS TO ES2015
+		// IMPORTS. They are intentionally written as such because minification
+		// will drop this unreachable block in production environments.
+		if ( 'production' !== process.env.NODE_ENV ) {
+			const chalk = require( 'chalk' );
+			let message = 'Invalid block parse for ';
+
+			// Chalk automatically detects its environment to determine whether
+			// ANSI codes are supported. In these environments we provide a nicely
+			// colored error message for invalid parse.
+			//
+			// TODO: Teach browsers to ANSI
+			if ( chalk.enabled ) {
+				const { diffLines } = require( 'diff' );
+				const diff = diffLines( actual, expected );
+				message += chalk.bold( blockType.name ) + ':\n\n';
+				message += chalk.bgGreen( 'Expected' ) + ' ' + chalk.bgRed( 'Actual' ) + '\n\n';
+
+				message += diff.reduce( ( result, part ) => {
+					let { value } = part;
+					if ( part.removed ) {
+						value = value.replace( /\n$/, '' );
+						value = value.replace( /^/gm, '- ' );
+						value = chalk.red( value );
+					} else if ( part.added ) {
+						value = value.replace( /\n$/, '' );
+						value = value.replace( /^/gm, '+ ' );
+						value = '\n' + value;
+						value = chalk.green( value );
+						value += '\n';
+					}
+
+					return result + value;
+				}, '' );
+			} else {
+				message += blockType.name + ':\n\n' +
+					'Expected:\n\n' + expected + '\n\n' +
+					'Actual:\n\n' + actual;
+			}
+
+			if ( 'test' === process.env.NODE_ENV ) {
+				throw new Error( message );
+			} else {
+				// eslint-disable-next-line no-console
+				console.warn( message );
+			}
+		}
```

Comment:
I think this code block could be structured better to achieve the following goals:

- Serve as a foundation for an eventual UI that explains the differences to the user.
- Specifically test the logic for generating the difference messages.

We probably do want this code in the production build eventually, so these changes would end up removing most of the environment-specific conditionals.  What would remain would be something like this:

```js
if ( 'test' === process.env.NODE_ENV ) {
    // Maybe we can even remove this, in favor of
    // `expect( console.warn ).toHaveBeenCalledWith( ... )`
    // in the tests below.
    throw new Error( message );
} else if ( 'production' !== process.env.NODE_ENV ) {
    console.warn( message );
}
```